drivers/can/can_mcux_flexcan.c

```diff
@@ -0,0 +1,604 @@
+/*
+ * Copyright (c) 2019 Vestas Wind Systems A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <atomic.h>
+#include <can.h>
+#include <clock_control.h>
+#include <device.h>
+#include <misc/byteorder.h>
+#include <fsl_flexcan.h>
+
+#define LOG_LEVEL CONFIG_CAN_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(can_mcux_flexcan);
+
+/*
+ * RX message buffers (filters) will take up the first N message
+ * buffers. The rest are available for TX use.
+ */
+#define MCUX_FLEXCAN_MAX_RX CONFIG_CAN_MAX_FILTER
+#define MCUX_FLEXCAN_MAX_TX \
+	(FSL_FEATURE_FLEXCAN_HAS_MESSAGE_BUFFER_MAX_NUMBERn(0) \
+	- MCUX_FLEXCAN_MAX_RX)
+
+/*
+ * Convert from RX message buffer index to allocated filter ID and
+ * vice versa.
+ */
+#define RX_MBIDX_TO_ALLOC_IDX(x) (x)
+#define ALLOC_IDX_TO_RXMB_IDX(x) (x)
+
+/*
+ * Convert from TX message buffer index to allocated TX ID and vice
+ * versa.
+ */
+#define TX_MBIDX_TO_ALLOC_IDX(x) (x - MCUX_FLEXCAN_MAX_RX)
+#define ALLOC_IDX_TO_TXMB_IDX(x) (x + MCUX_FLEXCAN_MAX_RX)
+
+/* Convert from back from FLEXCAN IDs to Zephyr CAN IDs. */
+#define FLEXCAN_ID_TO_ZCAN_ID_STD(id) \
+	((u32_t)((((u32_t)(id)) & CAN_ID_STD_MASK) >> CAN_ID_STD_SHIFT))
+#define FLEXCAN_ID_TO_ZCAN_ID_EXT(id) \
+	((u32_t)((((u32_t)(id)) & (CAN_ID_STD_MASK | CAN_ID_EXT_MASK)) \
+	>> CAN_ID_EXT_SHIFT))
+
+struct mcux_flexcan_config {
+	CAN_Type *base;
+	char *clock_name;
+	clock_control_subsys_t clock_subsys;
+	int clk_source;
+	u32_t bitrate;
+	u32_t sjw;
+	u32_t prop_seg;
+	u32_t phase_seg1;
+	u32_t phase_seg2;
+	void (*irq_config_func)(struct device *dev);
+};
+
+struct mcux_flexcan_rx_callback {
+	flexcan_rx_mb_config_t mb_config;
+	flexcan_frame_t frame;
+	can_rx_callback_t function;
+	void *arg;
+};
+
+struct mcux_flexcan_tx_callback {
+	struct k_sem done;
+	int status;
+	flexcan_frame_t frame;
+	can_tx_callback_t function;
+	void *arg;
+};
+
+struct mcux_flexcan_data {
+	flexcan_handle_t handle;
+
+	ATOMIC_DEFINE(rx_allocs, MCUX_FLEXCAN_MAX_RX);
+	struct k_mutex rx_mutex;
+	struct mcux_flexcan_rx_callback rx_cbs[MCUX_FLEXCAN_MAX_RX];
+
+	ATOMIC_DEFINE(tx_allocs, MCUX_FLEXCAN_MAX_TX);
+	struct k_sem tx_allocs_sem;
+	struct mcux_flexcan_tx_callback tx_cbs[MCUX_FLEXCAN_MAX_TX];
+};
+
+static int mcux_flexcan_configure(struct device *dev, enum can_mode mode,
+				  u32_t bitrate)
+{
+	const struct mcux_flexcan_config *config = dev->config->config_info;
+	flexcan_config_t flexcan_config;
+	struct device *clock_dev;
+	u32_t clock_freq;
+
+	/* TODO: support silent mode */
```

Comment:
@MaureenHelm Do you think it would be possible - at a later time - to get an update to the MCUXpresso SDK to allow setting the `LOM` bit through `flexcan_config_t` similar to the loopback setting? The `LOM` bit can only be written in Freeze mode...

Reply:
A new version of the MCUXpresso SDK was released last week. Can you have a look to see if this perhaps got added already, and if not, I can submit a ticket.

Reply:
@MaureenHelm Thanks. This functionality is not included in MCUXpresso SDK 2.6.0 either.